sockeye/rerank.py

```diff
@@ -53,72 +49,61 @@ def __init__(self, metric: str,
+        ranking = list(np.argsort(scores, kind='mergesort')[::-1])  # descending
+        reranked_hypotheses = self._sort_by_ranking(hypotheses, ranking)
         if self.return_score:
-            sorted_scores = [scores[i] for i in sorted_indexes]
-            return RerankOutput(hypotheses=sorted_hypotheses,
-                                scores=sorted_scores)
-        else:
-            return RerankOutput(hypotheses=sorted_hypotheses,
-                                scores=[])
-
-    def rerank_top1(self, hypotheses: List[str],
-                    reference: str) -> RerankOutput:
-        """
-        Reranks a set of hypotheses that belong to one single reference
-        translation and outputs the best hypothesis.
+            reranked_hypotheses['scores'] = [scores[i] for i in ranking]
+        return reranked_hypotheses
 
-        :param hypotheses: List of nbest translations.
-        :param reference: A single string with the actual reference translation.
-        :return: The single best hypothesis, possibly with its score.
-        """
-        scores = [self.scoring_function(hypothesis, reference) for hypothesis in hypotheses]
-        best_index = np.argmax(scores)  # type: int
+    @staticmethod
+    def _sort_by_ranking(hypotheses: Dict[str, Any], ranking: List[int]) -> Dict[str, Any]:
+        def ranksort(l):
+            return [l[i] for i in ranking]
 
-        if self.return_score:
-            return RerankOutput(hypotheses=[hypotheses[best_index]],
-                                scores=[scores[best_index]])
-        else:
-            return RerankOutput(hypotheses=[hypotheses[best_index]],
-                                scores=[])
+        return {key: ranksort(value) for key, value in hypotheses.items()}
 
 
 def rerank(args: argparse.Namespace):
     """
-    Reranks a list of hypotheses acoording to a sentence-level metric.
+    Reranks a list of hypotheses according to a sentence-level metric.
     Writes all output to STDOUT.
 
     :param args: Namespace object holding CLI arguments.
     """
-    reranker = Reranker(args.metric)
+    reranker = Reranker(args.metric, args.return_score)
 
     with utils.smart_open(args.reference) as reference, utils.smart_open(args.hypotheses) as hypotheses:
-        for reference_line, hypothesis_line in zip(reference, hypotheses):
-            reference_line = reference_line.strip()
-            hypotheses = json.loads(hypothesis_line)
-
-            utils.check_condition(len(hypotheses) > 1, "Reranking strictly needs more than 1 hypothesis.")
+        for i, (reference_line, hypothesis_line) in enumerate(zip(reference, hypotheses), 1):
+            reference = reference_line.strip()
+            # Expects a JSON object with keys containing at least 'translations',
+            # as returned by sockeye.translate's nbest output
+            hypotheses = json.loads(hypothesis_line.strip())
+            assert 'translations' in hypotheses
```

Comment:
Could you use `check_condition` and add a message when this fails?

Reply:
sure!